ttl/ttlworker/job_manager.go

```diff
@@ -0,0 +1,582 @@
+			if err != nil {
+				logutil.Logger(m.ctx).Warn("fail to resize delete workers", zap.Error(err))
+			}
+		case <-scheduleTicker:
+			m.rescheduleJobs(se, now)
+		}
+	}
+}
+
+func (m *JobManager) resizeScanWorkers(count int) error {
+	var err error
+	m.scanWorkers, err = m.resizeWorkers(m.scanWorkers, count, func() worker {
+		return newScanWorker(m.delCh, m.notifyStateCh, m.sessPool)
+	})
+	return err
+}
+
+func (m *JobManager) resizeDelWorkers(count int) error {
+	var err error
+	m.delWorkers, err = m.resizeWorkers(m.delWorkers, count, func() worker {
+		return newDeleteWorker(m.delCh, m.sessPool)
+	})
+	return err
+}
+
+func (m *JobManager) resizeWorkers(workers []worker, count int, factory func() worker) ([]worker, error) {
+	if count < len(workers) {
+		logutil.Logger(m.ctx).Info("shrink ttl worker", zap.Int("originalCount", len(workers)), zap.Int("newCount", count))
+
+		for _, w := range workers[count:] {
+			w.Stop()
+		}
+		var errs error
+		for _, w := range workers[count:] {
+			err := w.WaitStopped(m.ctx, 30*time.Second)
+			if err != nil {
+				logutil.Logger(m.ctx).Warn("fail to stop ttl worker", zap.Error(err))
+				errs = multierr.Append(errs, err)
+			}
+		}
+
+		// remove the existing workers, and keep the left workers
+		workers = workers[:count]
+		return workers, errs
+	}
+
+	if count > len(workers) {
+		logutil.Logger(m.ctx).Info("scale ttl worker", zap.Int("originalCount", len(workers)), zap.Int("newCount", count))
+
+		for i := len(workers); i < count; i++ {
+			w := factory()
+			w.Start()
+			workers = append(workers, w)
+		}
+		return workers, nil
+	}
+
+	return workers, nil
+}
+
+func (m *JobManager) updateTaskState() {
+	results := m.pollScanWorkerResults()
+	for _, result := range results {
+		job := findJobWithTableID(m.runningJobs, result.task.tbl.ID)
+		if job != nil {
+			logutil.Logger(m.ctx).Debug("scan task state changed", zap.String("jobID", job.id))
+
+			job.finishedScanTaskCounter += 1
+			job.scanTaskErr = multierr.Append(job.scanTaskErr, result.err)
+		}
+	}
+}
+
+func (m *JobManager) pollScanWorkerResults() []*ttlScanTaskExecResult {
+	results := []*ttlScanTaskExecResult{}
+	for _, w := range m.scanWorkers {
+		worker := w.(*ttlScanWorker)
+		result := worker.PollTaskResult()
+		if result != nil {
+			results = append(results, result)
+		}
+	}
+
+	return results
+}
+
+// checkNotOwnJob removes the job whose current job owner is not yourself
+func (m *JobManager) checkNotOwnJob(se session.Session) {
+	for _, job := range m.runningJobs {
+		timeoutJobCtx, cancel := context.WithTimeout(m.ctx, ttlInternalSQLTimeout)
+
+		tableStatus := m.tableStatusCache.Tables[job.tbl.ID]
+		if tableStatus == nil || tableStatus.CurrentJobOwnerID != m.id {
+			logutil.Logger(m.ctx).Info("job has been taken over by another node", zap.String("jobID", job.id), zap.String("statistics", job.statistics.String()))
+			m.removeJob(job)
+			err := job.Cancel(timeoutJobCtx, se)
```

Comment:
Maybe it is not necessary to set the job's state to cancelled. "Cancelled" means it will never resume in the future, but in this case, it should be taken over and resumed by other workers.

Reply:
Yes! Fixed